server/client.go

```diff
@@ -1112,7 +1112,7 @@ func (c *client) flushOutbound() bool {
 	}
 
 	// Check to see if we can reuse buffers.
-	if len(cnb) > 0 {
+	if len(cnb) > 0 && n >= int64(len(cnb[0])) {
```

Comment:
Seems len(cnb[0]) will be the length after the write since of the *v thing we saw no?

Reply:
No, cnb[0] will still be the original nb[0]. What matters here is that we can't use the *original* nb[0] unless it has been fully written to the socket, hence checking that n >= len(cnb[0]) or in other words, the length of original nb[0].
The issue was that in the case of a partial where nb[0] is not "removed", but instead updated with the left over, so nb[0]="original slice[consumed:]", then we would use cnb[0][:0] which in this case really meant nb[0][:0] because nb[0] slice was "updated" and not removed and replaced with nb[1], etc.. this meant that we would possibly overwrite the remaining of nb[0] slice when writing more than what was consumed from the first slice.

Reply:
I am not sure I am making much sense, but cnb[0] will be point to nb[0] if and only if nb[0] slice was a partial. If it was fully written, nb[0] would then be replaced with nb[1], etc.. or end up with len(nb) == 0 if all slices have been written.
So len(cnb[0]) is correct to use. It would be equivalent of snapshotting nb[0] before the write, but again, bench has showed me that it is faster the way it is coded that my original PR..

Reply:
cnb[0] will be the original slice which will be updated such that of the write is successful it will be 0 and if it was a partial and that was the only slice then it will be whatever is left. So if we had only 1 buffer and we write a partial but N written bytes is larger than the remaining bytes, this conditional would pass incorrectly no?

Reply:
Sorry, you are right. I have updated the test to cause the code to fail and have now updated the check. I capture the first slice original length before the write and use that for the check. Modified test passes and bench still ok.